### DISCUSSION TITLE
Heirarchical Tagging/Categorization and Filtering Support

### DISCUSSION THREAD

**Discussion by radh21301** [seanmorley15/AdventureLog#290] [Ideas]
This is a beautiful app. Thanks you for this!

I have been using Google my maps for a long time now, but then tried multiple other apps (none were selfhosted). Many of the items in this feature request are inspired from an app called called [Proxi](https://www.proxi.co/). They have an extremely nice GUI and feels very refreshing to use -> better than google my maps in my opinion for some things. But it is not self hosted, and their privacy policy does not inspire confidence, especially for something involving location tracking.
- Categorization/Tagging 🏷️ : 
  -  Right now, on the GUI to add an adventure, it lets us select 'Lodging' or 'Restaurant'. Which is great. But I think it becomes limited when different people plan their trips differently. Instead of having these things hard-coded, they can be set as tags groups and tags (basically hierarchical tags). Then the options become unlimited.
  - Type of place (Tag group):
    - Attractions (tag)
    - Restaurants (tag)
    - Museums (tag)
    - Parks (tag)  ...
  - ⭐ ratings (Tag group):
      - Awesome (tag)
      - Good (tag)
      - Bad (tag)
   - In this example, one can Choose a tag from each group. So, they can choose the type of place (say Restaurant), and then choose a rating (Awesome)
   - People can basically do whatever they want with it. I think this would be really useful.
   - Each tag can have its own icon -> makes for an a great viewing experience
- Filtering 🔍 : Ability to view a filtered list of the places from those tags would be helpful. Say, filter out only restaurants.
- Optional (Google maps link) 📍: Google maps is really helpful to get directions to that place, read reviews etc. So, an option to directly take us to google maps is handy.

*For some inspiration, here are some screenshots from the Proxi app.*

------------------------------**Categories and Tags** ------------------------------------
<img width="329" alt="image" src="https://github.com/user-attachments/assets/c8632f21-1264-4dc8-b1f2-cf571c023f07">

------------------------------------**GUI to add a point** ------------------------------------
<img width="326" alt="image" src="https://github.com/user-attachments/assets/8b0af74a-c370-4c08-8330-12ff7047b236">

------------------------------------**The Map Page** ------------------------------------
<img width="1150" alt="image" src="https://github.com/user-attachments/assets/2ad13b79-7ed6-4bfb-a5e6-e0b7a3affe81">

**Comment by seanmorley15** [comment]
WOW! These are some amazing ideas! Here is how I envision being able to implement these things. 

Right now, adventures are either planned or visited on the main page but collections allow for dining and lodging (and other trip specific things like travel and notes). I could see making planned/visited a toggle and Doing the category idea instead. Activity types could be renamed to tags (auto complete from existing tags is already implemented). This would give the user more control over the types of adventures and allow them to easily tag them. The sorting screen can get a list of the users unique categories to sort them and there will be some default ones, but the user can add more (with emoji as icon!).

I think this overall gives the user more flexibility and openness. Feel free to chime in with any more ideas!

Just for reference, here is how I envision the roadmap for AdventureLog:

 - Add sharing functionality for collections. The user can share a collection with other users so they can collaborate (started and API is implemented in feature branch)
 - Do this - Migrate to visited/planned toggle and rely more on tagging for things like activities and giving each adventure a category.

My free time is less now with school so I cannot do a new feature everyday like I did in the summer 😂 but I may be able to get a team together to help me make AdventureLog the best it can be :)
I will keep this post updated as I make progress!

**Comment by radh21301** [comment]
That sounds great!

Here are some things to think about -> Just suggestions.
* The toggle for planned/visited can also include an All option. Basically, see everything (maybe have different colors, or shading or something to distinguish them)
* Your thought is to use Categories and Tags (single level. meaning no subtags)
  * Example would be:
    * Category = Activity
    * Tag can be any of the following:
      * Swimming
      * Hiking
      * Snorkeling
      * Surfing
      * Awesome
      * Terrible ...
    * Basically, it would allow a user to select Hiking AND Surfing. That is not great, in my opinion. It is a type of activity, so only one should be chosen. This can be solved using Tag groups and Sub-Tags - as in the following example
* If we use Tag Groups and Sub-Tags. This is what it would look like. Note: A user can choose only one sub-tag in a group
  * Category = Activity
  * Type (This is a tag group. User can choose one of the following. or not choose anything)
      * Swimming
      * Hiking
      * Snorkeling
      * Surfing
  * Rating (Then choose one of the following if they want to)
    * Awesome
    * Terrible

The fact is, a user is not required to add a tag group. If they just want a simple tag, go ahead with those. But those who want more flexibility can choose to add tag groups.

This might sound more like a power user kind of feature, and maybe it is. But I think users would definitely get the hang of it once they start using it OR with examples on the website of how this can be used. I never used tagging for this sort of thing before Proxi app. But it definitely made things much much more manageable.

These are just some thoughts. Just take elements that fit into the vision you have for AdventureLog!

I have a collection of all the restaurants I have been to and the ones I plan to. And they are sorted by cuisine with their own icons. Looks pretty cool on a map.

**Comment by radh21301** [comment]
On second thought. Enforcing users to choose only one sub-tag from a tag group might be limiting and too strict (maybe unnecessary complication?). If they choose multiple sub-tags, they will realize that and probably remove them OR they actually want those multiple tags.. The responsibility can be left to the user.

I only see Tag-Groups helping when filtering things we want to see. The following looks much better and easier on the eyes than just having a single list.


|        With Tag Groups                      | Without Tag Groups                                        |
|---------------------------------------------------|----------------------------------------------------|
| - Type<br>   --- Swimming<br> --- Hiking<br> --- Snorkeling<br> --- Surfing<br> - Rating <br> --- Awesome<br> --- Terrible | - Hiking<br> - Snorkeling<br> - Surfing<br> - Awesome<br> - Terrible <br>  - Swimming |

**Note**: Swimming is in the end..after Rating (Awesome/Terrible). I am not sure how the tags will be sorted -> Alphabetically, Tag Creation Time/Order etc.. But using tag groups keeps similar things together. Makes it easier for the viewer to select what they what.

**Comment by seanmorley15** [comment]
Quick update to this, the initial version of this will include migration to categories instead of visited and planned. visited and planned will be determined by the number of visits recorded and a set of (for now) predefined categories will be used instead. Here is my current list, feel free to suggest any more
```python
ADVENTURE_TYPES = [
    ('general', 'General 🌍'),
    ('Outdoor', 'Outdoor 🏞️'),
    ('lodging', 'Lodging 🛌'),
    ('dining', 'Dining 🍽️'),
    ('activity', 'Activity 🏄'),
    ('attraction', 'Attraction 🎢'),
    ('shopping', 'Shopping 🛍️'),
    ('nightlife', 'Nightlife 🌃'),
    ('event', 'Event 🎉'),
    ('transportation', 'Transportation 🚗'),
    ('culture', 'Culture 🎭'),
    ('water_sports', 'Water Sports 🚤'),
    ('hiking', 'Hiking 🥾'),
    ('wildlife', 'Wildlife 🦒'),
    ('historical_sites', 'Historical Sites 🏛️'),
    ('music_concerts', 'Music & Concerts 🎶'),
    ('fitness', 'Fitness 🏋️'),
    ('art_museums', 'Art & Museums 🎨'),
    ('festivals', 'Festivals 🎪'),
    ('spiritual_journeys', 'Spiritual Journeys 🧘‍♀️'),
    ('volunteer_work', 'Volunteer Work 🤝'),
    ('other', 'Other')
]
```
This will be released as a part of the v0.7.0 update.

**Reply by radh21301** [reply]
This looks great! I only use 5/6 categories and all of those are in your list. The second line has ('Outdoors' , 'Outdoors') - both are in caps.

May I ask why you decided to go with built in categories rather than user defined ones for now?

**Reply by seanmorley15** [reply]
Hi @radh21301, I will definitely fix the minor bug mentioned. Regarding the built-in categories, I just am doing this for now so I can get the actual feature shipped. This is because allowing for custom categories right now adds a lot of extra overhead, so its just easier to make the feature then improve on it. I am making significant progress but this update is large due to the fact that it changed how the core actual adventure log feature works.

I will keep you updated!

**Reply by radh21301** [reply]
Wonderful! Thanks for all your work!
